Contributions to 'Feature/p file reader'

LiamPattinson opened the pull request:
Contributions to PR https://github.com/pyro-kinetics/pyrokinetics/pull/179

pFile reader requires an additional G-EQDSK file to determine its $\rho$ grid, so a `second_filename` arg was added to the `Reader` class. As this class is very generic and should be applicable throughout the project, I think it'd be better to keep it simple and make use of the `**kwargs` passed to the `read` function to handle any special cases. This PR reverts the changes to `Kinetics` and `Reader`, and instead passes `eq_file` via standard `**kwargs`. Further changes:

- Updated pFile reader ``verify`` method, automatic file type inference now working:
```python
>>> # No need for kinetics_type:
>>> import pyrokinetics as pk
>>> pyro = pk.Pyro(kinetics_file=pk.kinetics_templates["pFile"], eq_file=pk.eq_templates["GEQDSK"])
```
- ``Pyro`` now takes optional dicts `eq_kwargs` and `kinetics_kwargs` to its constructor, so complicated builds that previously required separate calls to `load_global_eq` and `load_global_kinetics` can now be done in one step.
- If ``Pyro`` has an ``eq_file`` and the user tries to load a pFile kinetics, it will detect if the error is simply a missing ``eq_file`` arg and correct it automatically.
- Updated Reader docs, removed unused `*args` and fixed some type hints

LiamPattinson committed:
Update to Reader, eq_file passed to pFile reader via kwargs

- Removed second_filename and *args from Reader, updated docs
- Reverted kinetics changes
- Updated pFile reader verify method, automatic filetype inferrence
  now working.
- eq_file now passed to pFile reader via regular kwargs, not using
  a special second_filename arg.
- Pyro now takes eq_kwargs and kinetics_kwargs to constructor
- If Pyro has an eq_file and the user tries to load a pFile kinetics,
  it will detect if the error is simply a missing eq_file arg and
  correct it

bpatel2107 commented:
This looks very nice! Can you add an example with `eq_kwargs` or `kinetic_kwargs`. You could also modify this script here

https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/pyrokinetics/examples/example_JETTO.py

Can you add a test for these kwargs? You could use `time` in TRANSP/JETTO netCDFs and `psi_n_lcfs` for GEQDSK

In #179 we can add some tests for when a `pFile` is loaded without a `eq_file`

bpatel2107 commented on pyrokinetics/kinetics/KineticsReaderpFile.py:
```diff
@@ -63,19 +64,30 @@ class KineticsReaderpFile(KineticsReader):
         """
-        # Read pFile, get generic data.
+        # eq_file must be provided
+        # Note: The 'Optional[T]' type hint just means the arg can be of type T or it
```
I'm not sure I understand this. Why wouldn't `eq_file: PathLike = None`  work?

LiamPattinson replied:
It'd work just fine, but a static type checker might complain, as the default value `None` isn't of type `PathLike`. The comment is just pointing out that `Optional[T]` is really an alias for `Union[T, None]` (or maybe `Union[T, type(None)]`?), and so despite the implications of the name 'optional', in this case we do actually require the user to provide something.

In other words, it's me being weirdly pedantic about type hints, so it's probably best to just remove the type hint and the comment since they're actually degrading readability in this case, and that kinda defeats the whole point.

LiamPattinson committed:
Add tests for eq/kinetics_kwargs

LiamPattinson committed:
Fixed copy-paste error in comments

LiamPattinson committed:
flake8 fix

LiamPattinson commented:
Sorry for rerunning CI repeatedly there, kept catching minor errors after pushing.

I changed the JETTO example to use `eq_kwargs` and `kinetics_kwargs`, and I added a couple of simple tests to check that kwargs were being forwarded properly.

As the pFile reader doesn't have tests yet, I think it'd be best to hold off on writing those tests until the PR is closer to being finished.

bpatel2107 reviewed (approved):
I think these simple tests are good enough for this. Happy to merge

LiamPattinson committed:
Edit to type hints and comments

jfparisi commented:
I can't find pk.kinetics_templates["pFile"]? Is this an error on my side, or do we need to upload one?

LiamPattinson commented:
It should already be there, as these lines added it:

https://github.com/pyro-kinetics/pyrokinetics/blob/ad3053a7019ca55e8f0b12bd21397f55480ca3de/pyrokinetics/templates.py#L24-L33

Did you definitely do `import pyrokinetics as pk` and not include any typos? If so, please could you post your error message?

Not sure if this is the issue, but to make sure you're running my changes, you'll need to clone a new pyrokinetics from `github.com/PlasmaFAIR/pyrokinetics` instead of `github.com/pyro-kinetics/pyrokinetics`, and then switch to the `feature/pFile_reader` branch. I started working on this project from a fork and still haven't broken the habit, and it only just occurred to me that this just makes it harder for others to test my code.

bpatel2107 commented:
I think you either need to checkout the forked PlasmaFAIR branch, or we'd need to merge this into your branch and then you can pull those changes to your local machine

bpatel2107 commented:
I'll just merge this so @jfparisi can pull the latest version of `feature/pFile_reader`

jfparisi commented:
@LiamPattinson @bpatel2107  I probably messed up. I was testing on: git fetch origin pull/180/head:feature/pFile_reader. Could you tell me the correct way to checkout these further updates and test from the command line?

LiamPattinson commented:
I'm honestly not sure if that would work -- I tend to use the same few basic git commands 99% of the time, and I don't think I've ever tried fetching/pulling from a PR. Now it's merged into your branch you should be able to just `git pull` on your regular branch, so that'll make things easier. When dealing with forks in general I'd normally handle it like this:

```bash
$ cd ~/my/workspace
$ # get the fork instead of the main version
$ git clone https://github.com/PlasmaFAIR/pyrokinetics pyrokinetics_fork
$ cd pyrokinetics_fork
$ # new python env (I use venv, conda works just as well)
$ python3 -m venv venv
$ source venv/bin/activate
$ # install, run tests
$ pip install --upgrade pip
$ pip install -e .[tests]
$ pytest
```
If you want to play about in the command line, it's usually best to do so from the top level directory, as relative import paths can break if you try to run things from within the source code itself.

If you're not happy with any of the merged changes, please let me know in the [original PR](https://github.com/pyro-kinetics/pyrokinetics/pull/179). We can always revert if we need to.

jfparisi commented:
Thanks for your help @LiamPattinson! The merged changes are excellent.

I am finding issues when writing GK input files. For example, with the below code, I get beta ~ 1e-25, vnewk ~ 1e-16 in the GK input files.
```
import pyrokinetics as pk
import pathlib
gs2_template = pk.template_dir / "input.gs2"
kinetics_file_in='/p/gyro/jparisi/equilibria/NSTX/nstx_132543/Oak_Dec2022/1.2_1.2_1.0/pFile'
eq_file_in = '/p/gyro/jparisi/equilibria/NSTX/nstx_132543/Oak_Dec2022/1.2_1.2_1.0/geqdsk'
base_path = '/p/gyro/jparisi/equilibria/NSTX/nstx_132543/Oak_Dec2022/1.2_1.2_1.0/'
pyro = pk.Pyro(kinetics_file=kinetics_file_in, eq_file=eq_file_in)
pyro.load_local(psi_n=0.5, local_geometry="Miller")
pyro.gk_code = "GS2"
base_path = pathlib.Path(base_path)
pyro.write_gk_file(file_name=base_path / "test_pFile.gs2")
pyro.write_gk_file(file_name=base_path / "test_pFile.cgyro", gk_code="CGYRO")
pyro.write_gk_file(file_name=base_path / "test_pFile.gene", gk_code="GENE")
pyro.write_gk_file(file_name=base_path / "test_pFile.tglf", gk_code="TGLF")
```
I have uploaded the geqdsk and pFile to[ this Google Drive folder](https://drive.google.com/drive/u/0/folders/1r6qsgUNy3r-S-NQn3E3n0xlkl4NBfYpY). Is this user error on my side?

bpatel2107 commented:
This sounds like a units issue. Can you check that the density and temper are in units of `/m3` and `eV`?

bpatel2107 commented:
Yeah that should work
